File: utilities/gl_gurobi.m
function [x, num_iters, out] = gl_gurobi(initial_x, A, b, mu, opts)
    % Solve the group LASSO problem using Gurobi through its MATLAB interface.
    %
    % Args:
    %   initial_x (matrix): Initial guess for the variable X
    %   A (matrix): Constraint matrix
    %   b (matrix): Observation vector
    %   mu (double): Regularization parameter
    %   opts (struct): Additional algorithm options (optional)
    
    if nargin < 5
        opts = struct();
    end
    
    [m, n] = size(A);
    num_targets = size(b, 2);
    
    % Initialize Gurobi model
    model = struct();
    model.modelsense = 'min';
    
    % Setup variables
    % X variables: n x num_targets matrix
    num_vars = n * num_targets + m * num_targets + n;
    model.lb = [-inf(n * num_targets, 1); -inf(m * num_targets, 1); zeros(n, 1)];
    model.ub = [inf(n * num_targets, 1); inf(m * num_targets, 1); inf(n, 1)];
    model.start = [reshape(initial_x, [], 1); zeros(m * num_targets + n, 1)];
    
    % Setup quadratic objective
    % Objective: 0.5 * ||Y||_F^2 + mu * sum(t)
    Q = sparse(num_vars, num_vars);
    Q(n*num_targets+1:n*num_targets+m*num_targets, n*num_targets+1:n*num_targets+m*num_targets) = 0.5 * speye(m*num_targets);
    model.Q = Q;
    c = zeros(num_vars, 1);
    c(end-n+1:end) = mu;
    model.obj = c;
    
    % Setup constraints
    % A * X[:, j] - b[:, j] == Y[:, j] for each target j
    Aeq = sparse(m * num_targets, num_vars);
    beq = reshape(b, [], 1);
    
    for j = 1:num_targets
        rows = (j-1)*m + (1:m);
        cols_x = (j-1)*n + (1:n);
        cols_y = n*num_targets + (j-1)*m + (1:m);
        Aeq(rows, cols_x) = A;
        Aeq(rows, cols_y) = -speye(m);
    end
    model.A = [Aeq; sparse(n, num_vars)];
    model.rhs = [beq; zeros(n, 1)];
    model.sense = [repmat('=', m*num_targets, 1); repmat('<', n, 1)];
    
    % Add SOC constraints: ||X[i, :]||_2 <= t[i]
    model.quadcon = struct('Qc', cell(1,n), 'q', cell(1,n), 'rhs', cell(1,n));
    for i = 1:n
        Qc = sparse(num_vars, num_vars);
        for k = 1:num_targets
            idx = (k-1)*n + i;
            Qc(idx, idx) = 1;
        end
        q = zeros(num_vars, 1);
        q(end-n+i) = -1;
        model.quadcon(i).Qc = Qc;
        model.quadcon(i).q = q;
        model.quadcon(i).rhs = 0;
    end
    
    % Set Gurobi parameters
    params.OutputFlag = 1;
    params.LogFile = utils.cvxLogsName;
    
    % Solve the model
    result = gurobi(model, params);
    
    % Read solver logs
    fid = fopen(utils.cvxLogsName, 'r');
    logs = fscanf(fid, '%c');
    fclose(fid);
    
    % Parse iterations from Gurobi logs
    iterations = utils.parse_iters(logs, 'GUROBI');
    
    % Print solver details and results
    fprintf('#######==Solver: GUROBI==#######\n');
    fprintf('Objective value: %f\n', result.objval);
    fprintf('Status: %s\n', result.status);
    fprintf('#######==Gurobi''s Logs:==#######\n%s\n', logs);
    fprintf('#######==END of Logs:==#######\n');
    fprintf('Parsed iterations:\n'); disp(iterations);
    
    % Check if iterations were recorded
    if isempty(iterations)
        fprintf('ERROR: Solver GUROBI recorded zero iterations. Please check stdout redirection!\n');
        num_iters = -1;
    else
        num_iters = size(iterations, 1);
    end
    
    % Prepare output structure
    out = struct();
    out.iters = iterations;
    out.fval = result.objval;
    
    % Extract solution
    x = reshape(result.x(1:n*num_targets), n, num_targets);
end

File: utilities/gl_gurobi.py
import gurobipy as gp
import numpy as np
import utilities.utils as utils

def gl_gurobi(initial_x: np.ndarray, A: np.ndarray, b: np.ndarray, mu: float, opts={}):
    """
    Solve the group LASSO problem using Gurobi through its Python interface.
    
    Args:
        initial_x (np.ndarray): Initial guess for the variable X.
        A (np.ndarray): Constraint matrix.
        b (np.ndarray): Observation vector.
        mu (float): Regularization parameter.
        opts (dict, optional): Additional algorithm options. Defaults to {}.
    
    Returns:
        tuple: Optimal solution X.x, number of iterations iters_N, and output dictionary out.
    """
    m, n = A.shape
    num_targets = b.shape[1]
    
    # Initialize Gurobi model
    model = gp.Model()
    
    # Add optimization variables
    X = model.addMVar((n, num_targets), lb=-gp.GRB.INFINITY, name="X")
    X.start = initial_x
    Y = model.addMVar((m, num_targets), lb=-gp.GRB.INFINITY, name="Y")
    t_variables = model.addMVar(n, lb=0.0, name="t")
    
    # Add constraints: A * X[:, j] - b[:, j] == Y[:, j] for each target j
    for j in range(num_targets):
        model.addConstr(A @ X[:, j] - b[:, j] == Y[:, j], name=f"Constraint_Y_{j}")
    
    # Add constraints: ||X[i, :]||_2 <= t[i] for each feature i
    for i in range(n):
        model.addConstr(gp.quicksum(X[i, k] * X[i, k] for k in range(num_targets)) <= t_variables[i] * t_variables[i],
                        name=f"Constraint_t_{i}")
    
    # Define the objective function: 0.5 * ||A*X - b||_F^2 + mu * sum(t)
    objective = 0.5 * gp.quicksum(Y[:, j].dot(Y[:, j]) for j in range(num_targets)) + mu * gp.quicksum(t_variables[i] for i in range(n))
    model.setObjective(objective, gp.GRB.MINIMIZE)
    
    # Optimize the model
    model.optimize()
    
    # Read the solver logs from the specified log file
    with open(utils.cvxLogsName, 'r', encoding='utf-8') as log_file:
        logs = log_file.read()
    
    # Parse the number of iterations from the logs specific to Gurobi
    iterations = utils.parse_iters(logs, 'GUROBI')
    
    # Logging solver details and results
    utils.logger.debug("#######==Solver: GUROBIPY==#######")
    utils.logger.debug(f"Objective value: {model.objVal}")
    utils.logger.debug(f"Status: {model.status}")
    utils.logger.debug(f"#######==Gurobi's Logs:==#######\n{logs}")
    utils.logger.debug(f"#######==END of Logs:==#######")
    utils.logger.debug(f"Parsed iterations:\n{iterations}")
    
    # Check if any iterations were recorded; if not, log an error
    if len(iterations) == 0:
        utils.logger.error("Solver GUROBI recorded zero iterations. Please check stdout redirection!")
        iters_N = -1
    else:
        iters_N = len(iterations)
    
    # Prepare the output dictionary with iterations and final objective value
    out = {
        'iters': iterations,          # List of tuples containing iteration number and objective value [(iter, fval), ...]
        'fval': model.objVal          # Final objective function value
    }
    
    # Return the optimal solution, number of iterations, and output dictionary
    return X.x, iters_N, out